### ISSUE TITLE
unexpected behaviour using indexing:depth

### ISSUE THREAD

**Issue by afoninsky** [mcollina/bloomrun#55]
```
const bloomrun = require('bloomrun')({
  indexing: 'depth'
})
bloomrun.add({ role: 'tag', cmd: 'find' }, 'tag,find')
bloomrun.add({ role: 'location', cmd: 'find' }, 'location,find')
bloomrun.add({ role: 'tag', cmd: 'find', count: /.*/ }, 'tag,find,count')
bloomrun.add({ role: 'location', cmd: 'find', count: /.*/ }, 'location,find,count')

const test = {}
test['tag,find,count'] = bloomrun.lookup({ role: 'tag', cmd: 'find', count: 'true' })
test['tag,find'] = bloomrun.lookup({ role: 'tag', cmd: 'find' })
test['location,find,count'] = bloomrun.lookup({ role: 'location', cmd: 'find', count: 'true'})
test['location,find'] = bloomrun.lookup({ role: 'location', cmd: 'find' })
console.log(test)
```

Expected output:
```
{ 'tag,find,count': 'tag,find,count',
  'tag,find': 'tag,find',
  'location,find,count': 'location,find,count',
  'location,find': 'location,find' }
```

Actual output:
```
{ 'tag,find,count': 'tag,find,count',
  'tag,find': 'tag,find',
  'location,find,count': 'location,find', <--- here
  'location,find': 'location,find' }
```

---
yarn --version
0.24.4

node -v
v7.8.0
--- yarn.lock content:
```
# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


bloomrun@^3.0.4:
  version "3.0.4"
  resolved "https://registry.yarnpkg.com/bloomrun/-/bloomrun-3.0.4.tgz#b408fad41186c144b717c2abe7684228582ec463"
  dependencies:
    sorted-array-functions "^1.0.0"

sorted-array-functions@^1.0.0:
  version "1.0.0"
  resolved "https://registry.yarnpkg.com/sorted-array-functions/-/sorted-array-functions-1.0.0.tgz#c0b554d9e709affcbe56d34c1b2514197fd38279"
```

**Comment by afoninsky** [comment]
Additional info: everithing is OK with such yarn.lock:
```
# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1

bloomrun:
  version "3.0.1"
  resolved "https://registry.yarnpkg.com/bloomrun/-/bloomrun-3.0.1.tgz#997732beb1c2b89ac163442befe4588120b0b87c"
  dependencies:
    sorted-array-functions "^1.0.0"

sorted-array-functions@^1.0.0:
  version "1.0.0"
  resolved "https://registry.yarnpkg.com/sorted-array-functions/-/sorted-array-functions-1.0.0.tgz#c0b554d9e709affcbe56d34c1b2514197fd38279"
```

**Comment by mcollina** [comment]
@afoninsky Can you please check if 3.0.2 and 3.0.3 have the same issue?

@gabssnake I think it's your #50.

**Comment by afoninsky** [comment]
3.0.2 - OK
3.0.3 - fail

**Comment by mcollina** [comment]
So, it's https://github.com/mcollina/bloomrun/pull/51. @gabssnake can you have a look?

alternatively, @afoninsky would you like to attempt a fix?

**Comment by afoninsky** [comment]
> alternatively, @afoninsky would you like to attempt a fix?
have no idea what's going on there, sorry :)

**Comment by StarpTech** [comment]
Is anybody working on this? @mcollina is a revert a solution?

**Comment by mcollina** [comment]
@StarpTech I'm not working on this.

I would prefer to avoid a revert but have a fix for both this and
https://github.com/mcollina/bloomrun/pull/51/files#diff-1dd241c4cd3fd1dd89c570cee98b79dd test.

**Comment by StarpTech** [comment]
I also don't work on this. @afoninsky @gabssnake feel free to investigate in it.

**Comment by StarpTech** [comment]
I found the issue but I'm working on a solution. If we change this line

https://github.com/mcollina/bloomrun/blob/489ef4df7ebfc7f9577e3a9ea1523404e378aa7b/lib/bucket.js#L16

to 

```js
sorted.add(this.data, set, this._algo)
```

it works but this doesn't solve the case `depth indexing preserves insertion order for the same pattern`

after some dig into it I found out the sort order of the bucket is wrong and that's why `{ "role": "location", "cmd": "find" }` (Weigth 2) is matched before `"role": "location", "cmd": "find", "count": /.*/` (Weigth: 3)

```json
[
  {
    "data": [
      {
        "pattern": { "role": "tag", "cmd": "find", "count": {} },
        "payload": "tag,find,count",
        "weight": 3
      },
      {
        "pattern": { "role": "tag", "cmd": "find" },
        "payload": "tag,find",
        "weight": 2
      },
      {
        "pattern": { "role": "location", "cmd": "find" },
        "payload": "location,find",
        "weight": 2
      },
      {
        "pattern": { "role": "location", "cmd": "find", "count": {} },
        "payload": "location,find,count",
        "weight": 3
      }
    ],
    "_isDeep": true,
    "weight": 3
  },
  {
    "data": [
      {
        "pattern": { "role": "location", "cmd": "find", "count": {} },
        "payload": "location,find,count",
        "weight": 3
      },
      {
        "pattern": { "role": "location", "cmd": "find" },
        "payload": "location,find",
        "weight": 2
      }
    ],
    "_isDeep": true,
    "weight": 3
  }
]
```

@mcollina for me it's unclear why we should support the feature of correct order of duplicate pattern when the strategy is based on the depth (weight) ? We should remove it and document that duplicate pattern are not ordered insertion order.

**Comment by mcollina** [comment]
I’m +1 with your proposal. I’ll do a major release for that change.

**Comment by StarpTech** [comment]
Nice 👍 I will create a PR tomorrow.

**Comment by mcollina** [comment]
Fixed in 4.0.0.
